File: app/src/main/java/ar/android/lflanzoni/norrisapp/Joke/jokeView.java
package ar.android.lflanzoni.norrisapp.Joke;

import android.content.Intent;
import android.os.PersistableBundle;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import ar.android.lflanzoni.norrisapp.Models.Jokes;
import ar.android.lflanzoni.norrisapp.R;

public class jokeView extends AppCompatActivity implements JokeMVP.View {

    JokeMVP.Presenter mPresenter;
    TextView tvJoke;
    ImageView ivJoke;
    String category;
    Boolean flagRestore;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.joke_view);
        mPresenter = new jokePresenter(this);
        tvJoke=  new TextView(this);
        tvJoke= findViewById(R.id.textViewJoke);
        ivJoke = new ImageView(this);
        ivJoke = findViewById(R.id.imgJoke);
        ivJoke.setImageResource(R.drawable.ic_norrisicono);
        if(savedInstanceState!=null){
            this.onRestoreInstanceState(savedInstanceState);
        }else {
            Intent i = this.getIntent();
            if (i.getStringExtra("category") != null) {
                category = i.getStringExtra("category");
            }

            mPresenter.requestDataFromServer(category);
        }
    }

    @Override
    public void setDataToTextView(Jokes randomJoke) {
        tvJoke.setText(String.valueOf(randomJoke.getValue()));
    }

    @Override
    public void onResponseFailure(Throwable throwable) {
        Toast.makeText(this,throwable.getMessage(),Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putCharSequence("joke",tvJoke.getText());

    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        tvJoke.setText(savedInstanceState.getCharSequence("joke"));

    }

    @Override
    protected void onResume() {
        super.onResume();
    }
}

File: app/src/main/java/ar/android/lflanzoni/norrisapp/joke/JokeView.kt
package ar.android.lflanzoni.norrisapp.joke

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import ar.android.lflanzoni.norrisapp.R
import ar.android.lflanzoni.norrisapp.Resource
import ar.android.lflanzoni.norrisapp.databinding.JokeViewBinding
import org.koin.android.viewmodel.ext.android.viewModel

class JokeView : AppCompatActivity() {

    private lateinit var binding: JokeViewBinding
    private val jokeViewModel: JokeViewModel  by viewModel()
    private lateinit var category: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = JokeViewBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.imgJoke.setImageResource(R.drawable.ic_norrisicono)

        if (intent.getStringExtra("category") != null) {
            category = intent.getStringExtra("category")
            jokeViewModel.getJokes(category).observe(this,{result ->
                when(result){
                    is Resource.Loading-> {
                        binding.imgJoke.visibility = View.GONE
                        binding.progressBarJoke.visibility = View.VISIBLE
                    }
                    is Resource.Success-> {
                        binding.progressBarJoke.visibility = View.GONE
                        binding.imgJoke.visibility = View.VISIBLE
                        binding.textViewJoke.text = result.data.value
                    }
                    is Resource.Failure->{
                        binding.imgJoke.visibility = View.GONE
                        binding.progressBarJoke.visibility = View.VISIBLE
                        showAlertDialog()
                    }
                }
            })
        }else{
            jokeViewModel.getRandomJoke().observe(this,{result ->
                when(result){
                    is Resource.Loading-> {
                        binding.imgJoke.visibility = View.GONE
                        binding.progressBarJoke.visibility = View.VISIBLE
                    }
                    is Resource.Success-> {
                        binding.progressBarJoke.visibility = View.GONE
                        binding.imgJoke.visibility = View.VISIBLE
                        binding.textViewJoke.text = result.data.value
                    }
                    is Resource.Failure->{
                        binding.imgJoke.visibility = View.GONE
                        binding.progressBarJoke.visibility = View.VISIBLE
                       showAlertDialog()
                    }
                }
            })
        }
    }

    private fun showAlertDialog(){
        val builder: AlertDialog.Builder = let {
            AlertDialog.Builder(it, R.style.MyDialogTheme)
        }
        builder.setMessage(R.string.dialog_message)
                .setTitle(R.string.dialog_title)
                .setPositiveButton(R.string.ok) { _, _ ->
                    finish()
                }
        val dialog: AlertDialog = builder.create()
        dialog.show()
    }

}